impexp-core/src/main/java/org/citydb/core/operation/importer/database/content/DBImplicitGeometry.java

```diff
@@ -27,213 +27,276 @@
-					psUpdateImplicitGeometry.setObject(3, importer.getDatabaseAdapter().getGeometryConverter().getDatabaseObject(geometryObject, batchConn));
-				else
-					psUpdateImplicitGeometry.setNull(3, importer.getDatabaseAdapter().getGeometryConverter().getNullGeometryType(),
-							importer.getDatabaseAdapter().getGeometryConverter().getNullGeometryTypeName());
+				if (result.getValue() != null) {
+					psUpdateImplicitGeometry.setObject(2, importer.getDatabaseAdapter()
+							.getGeometryConverter().getDatabaseObject(result.getValue(), connection));
+				} else {
+					psUpdateImplicitGeometry.setNull(2, nullGeometryType, nullGeometryTypeName);
+				}
 
 				psUpdateImplicitGeometry.addBatch();
 				++batchCounter;
 			}
 
 			if (batchCounter > 0) {
 				importer.executeBatch(TableEnum.IMPLICIT_GEOMETRY);
-				batchConn.commit();
-				batchCounter = 0;
+				connection.commit();
 			}
 
+			return implicitGeometryId;
 		} finally {
-			if (rs != null) {
-				try {
-					rs.close();
-				} catch (SQLException e) {
-					//
+			lockManager.releaseLock(gmlId != null ? gmlId : libraryObject);
+			lock.unlock();
+		}
+	}
+
+	private Map.Entry<Long, GeometryObject> importRelativeGeometry(AbstractGeometry relativeGeometry, ImplicitGeometry implicitGeometry) throws CityGMLImportException, SQLException {
+		long geometryId = 0;
+		GeometryObject geometryObject = null;
+
+		if (importer.isSurfaceGeometry(relativeGeometry)) {
+			geometryId = surfaceGeometryImporter.importImplicitGeometry(relativeGeometry);
+		} else if (importer.isPointOrLineGeometry(relativeGeometry)) {
+			geometryObject = geometryConverter.getPointOrCurveGeometry(relativeGeometry);
+		} else {
+			importer.logOrThrowUnsupportedGeometryMessage(implicitGeometry, relativeGeometry);
+		}
+
+		return new AbstractMap.SimpleEntry<>(geometryId, geometryObject);
+	}
+
+	private Map.Entry<Long, Boolean> lookupRelativeGeometry(String gmlId) throws SQLException {
+		long implicitGeometryId = importer.getObjectId(gmlId);
+		boolean isGlobal = false;
+
+		if (implicitGeometryId < 0 && hasGmlIdColumn) {
+			implicitGeometryId = importer.getObjectId("#globalGeometry#" + gmlId);
+			if (implicitGeometryId < 0) {
+				PreparedStatement ps = getOrCreateRelativeGeometryLookup();
+				ps.setString(1, gmlId);
+				try (ResultSet rs = ps.executeQuery()) {
+					if (rs.next()) {
+						implicitGeometryId = rs.getLong(1);
+						importer.putObjectId("#globalGeometry#" + gmlId, implicitGeometryId, MappingConstants.IMPLICIT_GEOMETRY_OBJECTCLASS_ID);
+					}
 				}
+			}
+
+			isGlobal = true;
+		}
 
-				rs = null;
+		return new AbstractMap.SimpleEntry<>(implicitGeometryId, isGlobal);
+	}
+
+	private Map.Entry<Long, Boolean> lookupLibraryObject(String libraryObject) throws SQLException {
+		String lookupKey = "#libraryObject#" + libraryObject;
+		long implicitGeometryId = importer.getObjectId(lookupKey);
+
+		if (implicitGeometryId < 0) {
+			PreparedStatement ps = getOrCreateRelativeGeometryLookup();
```

Comment:
shoud use `getOrCreateLibraryObjectLookup` instead?

Reply:
Of course! Thanks, fixed with 1064757eaacc0e4327ba661c1fc85dfed1f89922